frontend/packages/kubevirt-plugin/integration-tests/tests/models/baseVMBuilder.ts

```diff
@@ -0,0 +1,163 @@
+import { FlavorConfig, ProvisionSource, Disk, Network, CloudInitConfig } from '../types/types';
+import { BaseVMBuilderData } from '../types/vm';
+import { K8sKind } from '@console/internal/module/k8s';
+import { getRandStr } from '../utils/utils';
+
+export abstract class BaseVMBuilder<T extends BaseVMBuilderData> {
+  protected model: K8sKind;
+
+  protected data: T;
+
+  constructor(kind: K8sKind, builder?: BaseVMBuilder<T>) {
+    this.model = kind;
+    this.data = builder
+      ? builder.getData()
+      : ({ networks: [], disks: [], storageClassAttributes: {} } as any);
+  }
+
+  public getOSID() {
+    return OSIDLookup[this.data.os];
+  }
+
+  public generateName(id?: string) {
+    const finalId = id || getRandStr(5);
+    this.data.name = [
+      this.model.abbr.toLowerCase(),
+      this.data.namespace,
+      finalId,
+      this.data.flavor?.flavor,
+      this.getOSID(),
+    ]
+      .filter((a) => a)
+      .map((a) => a.toLowerCase().replace(/[^-a-zA-Z0-9]/g, ''))
+      .join('-');
+    return this;
+  }
+
+  public setName(name: string) {
+    this.data.name = name;
+    return this;
+  }
+
+  public setDescription(description: string) {
+    this.data.description = description;
+    return this;
+  }
+
+  public setNamespace(namespace: string) {
+    this.data.namespace = namespace;
+    return this;
+  }
+
+  public setTemplate(template: string) {
+    this.data.template = template;
+    return this;
+  }
+
+  public setFlavor(flavor: FlavorConfig) {
+    this.data.flavor = flavor;
+    return this;
+  }
+
+  public setWorkload(workload: Workload) {
+    this.data.workload = workload;
+    return this;
+  }
+
+  public setOS(os: OperatingSystem) {
+    this.data.os = os;
+    return this;
+  }
+
+  public setProvisionSource(provisionSource: ProvisionSource) {
+    this.data.provisionSource = provisionSource;
+    return this;
+  }
+
+  public setNetworks(networks: Network[]) {
+    this.data.networks = networks;
+    return this;
+  }
+
+  public setDisks(disks: Disk[]) {
+    this.data.disks = disks;
+    return this;
+  }
+
+  public setStorageAttributes(storageAttributes: { [k: string]: string }) {
+    this.data.storageClassAttributes = storageAttributes;
+    return this;
+  }
+
+  public setCloudInit(cloudInit: CloudInitConfig) {
+    this.data.cloudInit = cloudInit;
+    return this;
+  }
+
+  /**
+   * Sets attributes configured in passed builder instance.
+   * @param builder BaseVMBuilder<T>
+   */
+  public setBuilder(builder: BaseVMBuilder<T>) {
+    const data = _.cloneDeep(builder.getData());
+    Object.keys(data)
+      .filter((key) => data[key] !== undefined)
+      .forEach((key) => {
+        this.data[key] = data[key];
+      });
+    return this;
+  }
+
+  public appendBuilder(builder: BaseVMBuilder<T>) {
+    const customAppendKeys = new Set(['networks', 'disks', 'storageClassAttributes']);
+    const data = _.cloneDeep(builder.getData());
```

Comment:
I forgot what was here before and it looks fine now, but getData returns cloned data already so it should be enough.

Reply:
Right, fixed